zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,563 @@
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
  *
  * Author: Jakub Dabek <jakub.dabek@intel.com>
  */
 
 #include <sof/lib/regions_mm.h>
 
-struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
 
 /**
- * @brief Fill vm_heaps array with information from zephyr.
+ * @brief Initialize new heap
  *
- * Virtual memory regions calculated in zephyr are translated here
- * to a struct that will keep all information on current allocations
- * and virtual to physical mappings that are related to heaps.
- * System heap is not a part of this information. It only refers to
- * virtual first heaps.
- * Has to be initialized after calculations for regions is done in zephyr.
+ * Initializes new heap based on region that heap is to be assigned to
+ * and config.
+ * @param cfg Configuration structure with block structure for the heap
+ * @param memory_region_attribute A virtual region identifier
+ * @param span Flag deciding if heap can do contiguous allocation.
+ *
+ * @retval ptr to a new heap if successful.
+ * @retval NULL on creation failure.
  */
-static int virtual_heaps_init(const struct device *unused)
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+	int memory_region_attribute, bool span)
 {
-	ARG_UNUSED(unused);
-
+	uint32_t core_id = cpu_get_id();
 	struct sys_mm_drv_region *virtual_memory_regions =
 		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
 
-	for (size_t i = 0;
-		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
-		i++) {
-		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
+	if (vmh_get_heap_by_attribute(memory_region_attribute) != NULL)
+		return NULL;
+
+	struct virtual_memory_heap *new_heap = rzalloc(SOF_MEM_ZONE_RUNTIME, 0,
+		SOF_MEM_CAPS_RAM, sizeof(struct virtual_memory_heap));
+
+	if (new_heap == NULL)
+		return NULL;
+
+	list_init(&new_heap->node);
+	struct heap_config new_config = { 0 };
+
+	/* Initialize list that will keep all active allocations*/
+	list_init(&new_heap->active_allocations);
+
+	/* Search for matching attribute so we place heap on right
+	 * virtual region.
+	 * In case of core heaps regions we count theme from 0 to max
+	 * available cores
+	 */
+	if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP)	{
+		new_heap->virtual_region =
+			&virtual_memory_regions[core_id];
+	} else {
+		for (size_t i = 0;
+			i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
+			i++) {
+
+			if (memory_region_attribute == virtual_memory_regions[i].attr) {
+				new_heap->virtual_region = &virtual_memory_regions[i];
+				break;
+			}
+		}
+	}
+
+	/* If no config was specified we will use default one */
+	if (cfg == NULL) {
+		vmh_get_default_heap_config(
+			new_heap->virtual_region,
+			&new_config);
+		cfg = &new_config;
+	}
+
+	/* Check if configuration provided can fit into virtual regions memory */
+	uint32_t total_requested_size = 0;
+
+	for (size_t i = 0; i < MAXIMUM_BLOCK_SIZES; i++) {
```

Comment:
I don't think `size_t` is a correct type for a simple loop iterator. `size_t` is for actual sizes, usually in bytes